base/ltcmd.dtx

```diff
@@ -4747,7 +4754,14 @@
     in~an~expandable~command~after~an~optional~argument.
     \c_@@_ignore_def_tl
   }
-\msg_new:nnnn { cmd } { invalid-bang }
+\msg_new:nnnn { cmd } { invalid-bang-before-mandatory }
+  { Invalid~argument~prefix~'!'~in~#1. }
+  {
+    The~prefix~'!'~is~only~allowed~for~trailing~optional~arguments.~
+    You~tried~to~apply~it~to~an~optional~argument~before~mandatory~'#2'.
+    \c_@@_ignore_def_tl
+  }
+\msg_new:nnnn { cmd } { invalid-bang-on-mandatory }
   { Invalid~argument~prefix~'!'~in~#1. }
```

Comment:
I kinda changed my mind.

These two messages share the same `<text>` (`#3` of `\msg_new:nnnn`) and only differ in `<more text>` (`#4` of `\msg_new:nnnn`). Thus they can be combined.

Reply:
Done in https://github.com/latex3/latex2e/pull/1199/commits/6ce39c0b3a29d7712c4be08bc7f880d3edd4f6e8, along with rebasing on `develop`.